### ISSUE TITLE
[Feature Request] named tensor support for `tensordot`(tensor contract)

### ISSUE THREAD

**Issue by hzhangxyz** [pytorch/pytorch#41667]
## 🚀 Feature
`tensordot` use name to specify which rank to contract

## Motivation

when tensor rank is high, `tensordot`(contract) is not easy to specify the contract rank

## Pitch

eg: there is two rank-2 tensor a and b
`torch.tensordot(a, b, dims=(["first name of a", "second name of a"], ["second name of b", "first name of b"]))`

## Alternatives

currently, we need to wrap `tensordot`, analyze the tensor name, and then tensordot, and then recovery the names

**Comment by zou3519** [comment]
@hzhangxyz I agree this would be nice to have. We'd happily accept a pull request implementing this (feel free to tag me as reviewer)
